File: src/main/java/frc/robot/subsystems/Drivetrain.java
package frc.robot.subsystems;

import com.revrobotics.CANSparkMax;
import com.revrobotics.CANSparkMaxLowLevel;
import com.revrobotics.RelativeEncoder;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants;
import frc.robot.utils.Utils;

/**
 * <p>A subsystem representing the typical drivetrain of 5160 robots.</p>
 * <p>See <a href="https://docs.wpilib.org/en/stable/docs/software/hardware-apis/motors/wpi-drive-classes.html#drive-modes">here</a> for an explanation of the various drive modes.</p>
 */
public class Drivetrain extends SubsystemBase {
    private final CANSparkMax left1 = new CANSparkMax(Constants.drivetrainConstants.motors_left[0], CANSparkMaxLowLevel.MotorType.kBrushless);
    private final CANSparkMax left2 = new CANSparkMax(Constants.drivetrainConstants.motors_left[1], CANSparkMaxLowLevel.MotorType.kBrushless);
    private final CANSparkMax right1 = new CANSparkMax(Constants.drivetrainConstants.motors_right[0], CANSparkMaxLowLevel.MotorType.kBrushless);
    private final CANSparkMax right2 = new CANSparkMax(Constants.drivetrainConstants.motors_right[1], CANSparkMaxLowLevel.MotorType.kBrushless);

    private final MotorControllerGroup leftMotors = new MotorControllerGroup(left1, left2);
    private final MotorControllerGroup rightMotors = new MotorControllerGroup(right1, right2);

    private final DifferentialDrive differentialDrive = new DifferentialDrive(leftMotors, rightMotors);

    private final RelativeEncoder encoderLeft = left1.getEncoder();
    private final RelativeEncoder encoderRight = right1.getEncoder();

    public Drivetrain() {
        CANSparkMax[] motors = {left1, left2, right1, right2};
        for (CANSparkMax motor : motors) {
            configureSpark(motor);
        }

//        rightMotors.setInverted(true); // TODO: Uncomment if necessary
    }

    private void configureSpark(CANSparkMax motor) {
        motor.enableVoltageCompensation(Constants.drivetrainConstants.voltage_compensation);
        motor.setSmartCurrentLimit(Constants.drivetrainConstants.stall_current_limit);
        motor.setIdleMode(Constants.drivetrainConstants.default_mode);
        motor.setClosedLoopRampRate(Constants.drivetrainConstants.ramp_rate);
        motor.setInverted(Constants.drivetrainConstants.reverse_motor);
        motor.burnFlash();
    }

    /**
     * Drives using "tank controls", a system by which each side of the drivetrain is controlled independently.
     * @param leftPower the power of the left side of the drivetrain (from [-1..1]).
     * @param rightPower the power of the right side of the drivetrain (from [-1..1]).
     */
    public void tankDrive(final double leftPower, final double rightPower) {
        differentialDrive.tankDrive(leftPower, rightPower, false);
    }

    /**
     * Drives the robot at a certain power forward and with a certain amount of rotation.
     * @param power the power with which to drive forward (from [-1..1]).
     * @param rotation the power with which to rotate (proportional to the angular velocity, or how quickly the heading changes). (Must be from [-1..1]).
     */
    public void arcadeDrive(final double power, final double rotation) {
        differentialDrive.arcadeDrive(power, rotation, false);
    }

    /**
     * Drives the robot at a certain power forward and with a certain amount of steering.
     * This method makes turning easier at high speeds.
     * @param power the power with which to drive forward (from [-1..1]).
     * @param steering the amount of steering (inversely proportional to the turn radius).
     *                 Changing this value is can be thought of as changing how far a car's steering wheel is turned.
     *                 (Must be from [-1..1]).
     */
    public void curvatureDrive(final double power, final double steering) {
        differentialDrive.curvatureDrive(
            power,
            steering,
            true // TODO: What happens if this is false?
        );
    }

    /**
     * Stops the robot.
     */
    public void stop() {
        differentialDrive.stopMotor();
    }

    public double getAngularPosition() {
        double[] positions = { encoderLeft.getPosition(), encoderRight.getPosition() };
        double rotations = Utils.average(positions);
        double rotations_adjusted = rotations* Constants.drivetrainConstants.gear_ratio;
        double radians = rotations_adjusted*2*Math.PI;

        return radians;
    }

    public double getAngularVelocity() {
        double[] velocities = { encoderLeft.getVelocity(), encoderRight.getVelocity() };
        double rpm = Utils.average(velocities);
        double rpm_adjusted = rpm * Constants.drivetrainConstants.gear_ratio;
        double omega = (2*rpm_adjusted*Math.PI)/60;

        return omega;
    }

    public double getPosition() {
        double rotations = getAngularPosition();
        double distance = rotations*(Constants.drivetrainConstants.wheel_diameter/2);
        return distance;
    }

    public double getVelocity() {
        double radsec = getAngularVelocity();
        double vel = radsec*(Constants.drivetrainConstants.wheel_diameter/2);
        return vel;
    }
}

File: src/main/java/frc/robot/subsystems/Drivetrain.kt
@file:Suppress("Unused", "MemberVisibilityCanBePrivate")

package frc.robot.subsystems

import com.revrobotics.RelativeEncoder
import edu.wpi.first.wpilibj2.command.SubsystemBase
import frc.robot.Constants
import frc.robot.SuperPIDController
import frc.robot.SuperPIDControllerGroup
import kotlin.math.PI


// Config
//private const val turnPidKey = "Drivetrain Turn PID"
//private val turnPidConstants = PIDConstants(-0.1, 0.0, -0.01)
//
//private const val positionPidKey = "Drivetrain Position PID"
//private val positionPidConstants = PIDConstants(0.035, 0.0, 0.001)
//
private val autoPowerRange = -0.4..0.4
//private const val gearRatio = 1 / 10.71
//private const val wheelDiameter = 6.0
//private const val turnTolerance = 1.0
//private const val positionTolerance = 1.0

class Drivetrain(
    private val powertrain: Powertrain,
    private val navx: NavX
) : SubsystemBase() {
    @JvmField
    var state: State? = null
    private var power = 0.0
    private var rotation = 0.0

    private var turnPid: SuperPIDController = SuperPIDController(
        pidConstants = Constants.Drivetrain.turnPid,
        input = { distance },
        output = { rotation: Double ->
            this.rotation = rotation
            powertrain.mode = Powertrain.Mode.ARCADE_DRIVE
        },
        outputRange = autoPowerRange,
        tolerance = Constants.Drivetrain.turnTolerance,
        dashPidKey = Constants.Drivetrain.turnPidKey
    )
    
    private val positionPid: SuperPIDController = SuperPIDController(
        pidConstants = Constants.Drivetrain.positionPid,
        input = { distance },
        output = { rotation: Double ->
            power = rotation
            powertrain.mode = Powertrain.Mode.ARCADE_DRIVE
        },
        outputRange = autoPowerRange,
        tolerance = Constants.Drivetrain.positionTolerance,
        dashPidKey = Constants.Drivetrain.positionPidKey
    )
    private val pidControllerGroup: SuperPIDControllerGroup = SuperPIDControllerGroup(turnPid, positionPid)

    private val encoderLeft: RelativeEncoder = powertrain.left1.encoder
    private val encoderRight: RelativeEncoder = powertrain.right1.encoder
    private val encoders = listOf(encoderLeft, encoderRight)
    
    fun reset() {
        state = State.MANUAL_DRIVE
        encoderLeft.position = 0.0
        encoderRight.position = 1.0
    }

    init {
        reset()
    }

    val position: Double get() =
        encoders
            .map { it.position }
            .average()
            .adjustedForGearRatio *
                2 * PI // Convert to radians from rotations

    private val RelativeEncoder.wheelPosition get() = position * Constants.Drivetrain.gearRatio

    val rotationalVelocity: Double get() =
        encoders
            .map { it.velocity }
            .average()
            .adjustedForGearRatio *
                2 * PI / 60 // Convert to radians/sec from rotations/min

    private val Double.adjustedForGearRatio get() = this * Constants.Drivetrain.gearRatio

    val distance: Double get() = rotation * wheelRadius
    val velocity: Double get() = rotationalVelocity * wheelRadius

    private val wheelRadius: Double get() = Constants.Drivetrain.wheelDiameter / 2

    fun tankDrive(leftPower: Double, rightPower: Double) {
        pidControllerGroup.stopAll()
        state = State.MANUAL_DRIVE
        powertrain.tankDrive(leftPower, rightPower)
    }

    fun curvatureDrive(power: Double, rotation: Double) {
        pidControllerGroup.stopAll()
        state = State.MANUAL_DRIVE
        powertrain.curvatureDrive(power, rotation)
    }

    fun driveStraight(power: Double) {
        if (state != State.AIDED_DRIVE_STRAIGHT) {
            pidControllerGroup.stopAll()
            state = State.AIDED_DRIVE_STRAIGHT
            powertrain.mode = Powertrain.Mode.ARCADE_DRIVE
            turnPid.target = navx.heading
        }
        this.power = power
    }

    fun driveToPosition(position: Double) {
        if (state != State.PID_STRAIGHT) {
            pidControllerGroup.stopAll()
            powertrain.reset()
            state = State.PID_STRAIGHT
            positionPid.target = position + distance
        }
    }

    fun setPowerScaling(newScaling: Double) {
        powertrain.differentialDrive.setMaxOutput(newScaling)
    }

    fun stop() {
        state = State.STOPPED
        tankDrive(0.0, 0.0)
    }

    override fun periodic() {
        pidControllerGroup.executeAll()
        powertrain.periodic()
    }

    enum class State(val value: Int) {
        // 0-9: Manual Modes
        MANUAL_DRIVE(0),
        // 10-20: Aided Modes
        AIDED_DRIVE_STRAIGHT(10),
        STOPPED(11),
        // 20-29: PID Modes
        PID_TURNING(20),
        PID_STRAIGHT(21),
        PID_LIMELIGHT_TURNING(22),
        PID_LIMELIGHT_DRIVE(23);
    }
}